### PR TITLE
[Phi]Improve the mechanism for mkldnn kernel in PHI

### PR DESCRIPTION
**Pull request by YuanRisheng**
### PR types

Others
### PR changes

Others
### Describe

**Background:**
Paddle has a new Library named PHI. All Operator Kernels in Fluid will be moved to PHI in the future. Now, the MKLDNN Kernels can't be moved to PHI because it lacks the necessary mechanism. This PR improves the mechanism for mkldnn kernel in PHI.

**This PR's Work:**
1, Improve the compilation mechanism for mkldnn kernel in PHI
2, Improve the mkldnn kernel's register and selection mechanism in PHI
3, Some basic component related to mkldnn are moved to PHI
4, Move log_softmax mkldnn kernel to PHI

该PR主要是对phi库下MKLDNN kernel有关机制进行完善，主要做了以下工作：
1，完善对mkldnn kernel的编译，支持自动编译mkldnn kernel。
2，支持对mkldnn kernel的注册以及kernel选择。
3，mkldnn部分基础组件迁移到phi下。
4，迁移log_softmax的mkldnn kernel到phi下

### CODE REVIEW

File: paddle/phi/backends/onednn/onednn_context.cc [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifdef PADDLE_WITH_MKLDNN
+#include "paddle/phi/backends/onednn/onednn_context.h"
+
+#include "paddle/phi/api/ext/exception.h"
```
Comment (chenwhql) [inline paddle/phi/backends/onednn/onednn_context.cc]: Do we need this header file? if needed, replace it by enforce.h?

File: paddle/phi/backends/onednn/onednn_context.cc [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+// When Thread finish we clear oneDNN cache
+// This is needed when we have one executor used by many threads
+// e.g. test_analyzer_detect. Thread ID is not part of caching key
+// (for naive executor) so we need to clear cache when one thread finish
+// and other is to start inference
+// TODO(jczaja): Ideally it would be good to clear only part of cache
+// related to thread that is to be terminated
+MKLDNNContextThreadLocals::Body::~Body() {
+  auto cpu_place = phi::CPUPlace();
+  paddle::platform::DeviceContextPool& pool =
```
Comment (chenwhql) [inline paddle/phi/backends/onednn/onednn_context.cc:44]: TODO: we need remove the dependency on fluid device context here

File: paddle/phi/backends/onednn/onednn_context.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  static constexpr size_t kMKLDNNSessionID_Default = 0;
+  // mkldnn session id for cache clearing mode
+  static constexpr size_t kMKLDNNSessionID_CacheClearing = -1;
+  static Body& fetch() {
+    thread_local Body b;
+    return b;
+  }
+};
+
+class MKLDNNContext : public CPUContext {
```
Comment (chenwhql) [inline paddle/phi/backends/onednn/onednn_context.h]: Is it better to use OneDNNContext directly, consistent with the file name?

File: paddle/phi/kernels/onednn/log_softmax_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  astream.wait();
+
+  out->set_mem_desc(dst_memory_p->get_desc());
+}
+
+}  // namespace phi
+
+PD_REGISTER_KERNEL(log_softmax,
+                   MKLDNN,
+                   ALL_LAYOUT,
```
Comment (chenwhql) [inline paddle/phi/kernels/onednn/log_softmax_kernel.cc:69]: TODO: The layout of mkldnn kernel should be MKLDNN, we should support specifying the exact layout when the kernel is registered

File: paddle/phi/backends/onednn/onednn_context.cc (hunk 1 above)
Comment (YuanRisheng) [reply paddle/phi/backends/onednn/onednn_context.cc]: Has replaced it, thank you very much

File: paddle/phi/backends/onednn/onednn_context.h (hunk 3 above)
Comment (YuanRisheng) [reply paddle/phi/backends/onednn/onednn_context.h]: Has renamed it, thank you very much

File: paddle/phi/kernels/onednn/log_softmax_kernel.cc (hunk 4 above)
Comment (YuanRisheng) [reply paddle/phi/kernels/onednn/log_softmax_kernel.cc:69]: ok

File: paddle/phi/backends/onednn/onednn_context.cc (hunk 2 above)
Comment (YuanRisheng) [reply paddle/phi/backends/onednn/onednn_context.cc:44]: ok

**Review by jczaja — DISMISSED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]
LGTM

**Comment by YuanRisheng** [comment]
Please review my code, thank you! @jczaja @lidanqing-intel

**Review by jczaja — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**